include/dali/core/mm/memory.h

```diff
@@ -0,0 +1,153 @@
+  void operator()(void *memory) noexcept {
+    mr->deallocate(memory, size, alignment);
+  }
+};
+
+template <memory_kind kind>
+struct mr_stream_deallocator {
+  async_memory_resource<kind> *mr;
+  size_t size, alignment;
+  cudaStream_t release_on_stream;
+  void operator()(void *memory) noexcept {
+    if (release_on_stream == host_sync)
+      mr->deallocate(memory, size, alignment);
+    else
+      mr->deallocate_async(memory, size, alignment, release_on_stream);
+  }
+};
+
+template <typename T, memory_kind kind, typename Context>
+using mr_unique_ptr = std::unique_ptr<T, mr_deallocator<kind, Context>>;
+
+template <typename T, memory_kind kind>
+using mr_stream_unique_ptr = std::unique_ptr<T, mr_stream_deallocator<kind>>;
+
+template <typename T, memory_kind kind>
+void set_dealloc_stream(mr_stream_unique_ptr<T, kind> &ptr, cudaStream_t stream) {
+  ptr.get_deleter().release_on_stream = stream;
+}
+
+/**
+ * @brief Allocates uninitialized storage for `count` objects of type `T`
+ *
+ * This function allocates raw memory with suitable size and alignment to accommodate `count`
+ * object of type `T`. The memory is obtained from memory resource `mr`.
+ * The size of memory is at least sizeof(T) * count and the alignment is at least alignof(T).
+ * The return value is a unique pointer with a deleter that will safely dispose of the allocated
+ * buffer, but does not call destructors of the objects.
+ *
+ * @param mr        Memory resources to allocate the memory from.
+ * @param count     Number of objects for which the storage should suffice.
+ * @tparam T        Type of the object for which the storage is allocated.
+ * @tparam kind     The kind of requested memory.
+ * @tparam Context  The execution context in which the memory will be available.
+ */
+template <typename T, memory_kind kind, typename Context>
+mr_unique_ptr<T, kind, Context> alloc_raw(memory_resource<kind, Context> *mr, size_t count) {
+  size_t bytes = sizeof(T) * count;
+  size_t alignment = alignof(T);
+  void *mem = mr->allocate(bytes, alignment);
+  return mr_unique_ptr<T, kind, Context>(static_cast<T*>(mem),
+            mr_deallocator<kind, Context>{mr, bytes, alignment});
+}
+
+
+/**
+ * @brief Allocates uninitialized storage for `count` objects of type `T`
+ *
+ * This function allocates raw memory with suitable size and alignment to accommodate `count`
+ * object of type `T`. The memory is obtained from default memory resource.
+ * The size of memory is at least sizeof(T) * count and the alignment is at least alignof(T).
+ * The return value is a unique pointer with a deleter that will safely dispose of the allocated
+ * buffer, but does not call destructors of the objects.
+ *
+ * @param count     Number of objects for which the storage should suffice.
+ * @tparam T        Type of the object for which the storage is allocated.
+ * @tparam kind     The kind of requested memory.
+ * @tparam Context  The execution context in which the memory will be available.
```

Comment:
I don't think this param is available here.

Reply:
It's a template argument of the memory resource. It's not explicitly specified by the user, but it is there.

Reply:
Missed that.